text/0018-open-context-provider-framework.md

```diff
@@ -0,0 +1,157 @@
+---
+feature name: Open context provider framwork
+start date: 2020-05-27
+rfc pr: 
+related issue: https://github.com/aws/aws-cdk-rfcs/issues/18
+---
+
+# Summary
+
+The CDK has a super-handy feature in the form of the ContextProvider ( https://docs.aws.amazon.com/cdk/latest/guide/context.html ). These
+ContextProviders are currently hard-coded into the CDK CLI, and additional ContextProviders cannot be added by external sources. This
+proposal is to make the ContextProvider mechanism pluggable so that 3rd parties can create ContextProviders without adding them to the CDK
+repository, and users of the CDK can use these 3rd party ContextProviders.
+
+# Motivation
+
+Development within the CDK Construct library itself benefits greatly from this capability. When a new ContextProvider is needed, then a CDK
+contributor can just create it. Unfortunately, since the providers are hard-coded into the CDK, no 3rd party construct library can create their
+own ContextProvider; other than by contributing it to the CDK, of course. Some ContextProviders may be too specialized for reasonable consideration
+as a CDK contribution.
+
+Instead 3rd party libraries are required to implement workarounds, like CfnCustomResources that resolve during deployment instead of during
+synthesis. This increases the size of the Cfn template generated by the app, increases the time required to deploy the stack, adds unnecessary
+extra resources to the user's account, and makes it possible for a value generated by the CfnCustomResource to differ from one deployment
+to the next.
+
+Unfortunately, these CfnCustomResources required by 3rd party construct libraries may also be a security risk to the user. The lambda that
+backs them persists as long as the stack is deployed, and that lambda is granted permissions into the account that may be possible to leverage
+by an attacker that gains access to the lambda; depending on the lambda's purpose, of course.
+
+# Basic Example
+
+```ts
+// Allow the provider to be an arbitrary string
+const someValues: thirdpartlib.DoSomethingContextReponse = ContextProvider.getValue(scope, {
```

Comment:
I think I would want more protection around this than simply using `ContextProvider.getValue()`. At least there will need to be an agreement of a schema and checking on it (potentially using jsonschema, or maybe something simpler that doesn't bring too many dependencies).

If we're going to make an open-ended system, we want clear error detection and reporting.

Reply:
Also I think I would never require users to call `ContextProvider.getValue()`. Maybe construct library *authors* can, but users should have a stronger-typed interface on top.

They would be calling `ThirdPartyProvider.fromLookup()` or something.